tools/cfn2ts/lib/codegen.ts

```diff
@@ -593,20 +593,71 @@ export default class CodeGenerator {
           `${this.deeperCopy(arg).visitAtom(itemType)})`;
       }
 
-      public visitAtomUnion(_types: genspec.CodeName[]): string {
-        return this.baseExpression;
+      public visitAtomUnion(types: genspec.CodeName[]): string {
+        const validatorNames = types.map(type => genspec.validatorName(type).fqn).join(', ');
+
+        const arg = `prop${this.depth}`;
+        const mappers = types
+          .map(type => `(${arg}: any) => ${this.deeperCopy(arg).visitAtom(type)}`)
+          .join(', ');
+
+        return `${CFN_PARSE}.FromCloudFormation.getTypeUnion([${validatorNames}], [${mappers}], ${this.baseExpression})`;
       }
 
-      public visitListOrAtom(_scalarTypes: genspec.CodeName[], _itemTypes: genspec.CodeName[]): any {
-        return this.baseExpression;
+      public visitUnionList(itemTypes: genspec.CodeName[]): string {
+        const validatorNames = itemTypes.map(type => genspec.validatorName(type).fqn).join(', ');
+
+        const arg1 = `prop${this.depth}`;
+        const arg2 = `prop${this.depth + 1}`;
+        const mappers = itemTypes
+          // we need a copy 2 levels deep, as we're already using prop2 in arg2 here
+          .map(type => `(${arg2}: any) => ${this.deeperCopy(arg1).deeperCopy(arg2).visitAtom(type)}`)
```

Comment:
I had to spent a solid 20 minutes perusing the code to see why you needed 2 copies here. Let's not? :)

Reply:
Let me know if it's clearer why that's needed in this new version.